src/tlo/analysis/life_expectancy.py

```diff
@@ -0,0 +1,305 @@
+
+        # person-years lived in interval
+        py_lived_in_interval = pd.Series(index=range(number_age_groups), dtype=float)
+        for i in range(0, number_age_groups - 1):
+            py_lived_in_interval[i] = interval_width[i] * (
+                number_alive_at_start_of_interval[i + 1] + fraction_of_last_age_survived * number_dying_in_interval[i])
+        py_lived_in_interval[number_age_groups - 1] = number_alive_at_start_of_interval[number_age_groups - 1] / \
+                                                      death_rate_in_interval[number_age_groups - 1]
+
+        # person-years lived beyond start of interval
+        # have to iterate backwards for this
+        py_lived_beyond_start_of_interval = pd.Series(index=range(number_age_groups), dtype=float)
+        py_lived_beyond_start_of_interval[number_age_groups - 1] = py_lived_in_interval[number_age_groups - 1]
+        for i in range((number_age_groups - 2), -1, -1):
+            py_lived_beyond_start_of_interval[i] = py_lived_beyond_start_of_interval[i + 1] + py_lived_in_interval[i]
+
+        # calculate observed life expectancy at start of interval
+        condition = number_alive_at_start_of_interval == 0
+        observed_life_expectancy = pd.Series(index=range(number_age_groups), dtype=float)
+        observed_life_expectancy[condition] = 0
+        observed_life_expectancy[~condition] = py_lived_beyond_start_of_interval / number_alive_at_start_of_interval
+
+        # estimated life expectancy from birth
+        estimated_life_expectancy = estimated_life_expectancy.append(pd.Series([observed_life_expectancy[0]]))
+
+    return estimated_life_expectancy
+
+
+# generate life expectancy estimates for all draws/runs
+def produce_life_expectancy_estimates(results_folder: Path, median=True):
+    """
+    produces sets of life expectancy estimates for each draw/run
+    calls:
+    *1 num_deaths_by_age_group
+    *2 aggregate_person_years_by_age
+
+    Args:
+    - results_folder (PosixPath): The path to the folder containing all logged outputs
+    - median: declare whether to return a median value with 95% uncertainty intervals
+    or return the estimate for each draw/run
+
+    Returns:
+    - pd.DataFrame: The DataFrame with the life expectancy estimates (in years)
+    for every draw/run in the results folder
+    """
+    info = get_scenario_info(results_folder)
+
+    output = pd.DataFrame()
+
+    # extract numbers of deaths
+    deaths = _num_deaths_by_age_group(results_folder)
+
+    # extract person-years
+    person_years = _get_population_size_by_age(results_folder)
+
+    for draw in range(info['number_of_draws']):
+        for run in range(info['runs_per_draw']):
+            # select column
+            le = _estimate_life_expectancy(
+                number_of_deaths_in_interval=deaths.loc[:, (draw, run)],
+                person_years_at_risk=person_years.loc[:, (draw, run)])
+
+            output = pd.concat([output, le], axis=1)
```

Comment:
small point, but pandas is much faster if you collect these columns into a list, and then do `pd.concat` once on that  list. (I expect speed isn't an issue for this one, but just saying in case you see a situation that is more time-critical).

Reply:
Thanks Tim - I've made this change